File: PersianCalendar/src/main/java/com/byagowi/persiancalendar/view/preferences/AthanVolumeDialog.java
package com.byagowi.persiancalendar.view.preferences;

import android.content.Context;
import android.media.AudioManager;
import android.media.MediaPlayer;
import android.media.Ringtone;
import android.media.RingtoneManager;
import android.net.Uri;
import android.view.View;
import android.widget.SeekBar;

import com.byagowi.persiancalendar.R;
import com.byagowi.persiancalendar.util.UIUtils;
import com.byagowi.persiancalendar.util.Utils;

import java.io.IOException;

import androidx.preference.PreferenceDialogFragmentCompat;

public class AthanVolumeDialog extends PreferenceDialogFragmentCompat {
    private final String TAG = AthanVolumeDialog.class.getName();

    private int volume;
    private AudioManager audioManager;
    private Ringtone ringtone;
    private MediaPlayer mediaPlayer;

    @Override
    protected View onCreateDialogView(Context context) {
        View view = super.onCreateDialogView(context);

        final AthanVolumePreference athanPref = (AthanVolumePreference) getPreference();

        audioManager = (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);
        if (audioManager != null) {
            audioManager.setStreamVolume(AudioManager.STREAM_ALARM, athanPref.getVolume(), 0);
        }

        Uri customAthanUri = Utils.getCustomAthanUri(context);
        if (customAthanUri != null) {
            ringtone = RingtoneManager.getRingtone(context, customAthanUri);
            ringtone.setStreamType(AudioManager.STREAM_ALARM);
            ringtone.play();
        } else {
            MediaPlayer player = new MediaPlayer();
            try {
                player.setDataSource(context, UIUtils.getDefaultAthanUri(context));
                player.setAudioStreamType(AudioManager.STREAM_ALARM);
                player.prepare();
            } catch (IOException e) {
                e.printStackTrace();
            }
            try {
                player.start();
                mediaPlayer = player;
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        SeekBar seekBar = view.findViewById(R.id.sbVolumeSlider);

        volume = athanPref.getVolume();
        seekBar.setProgress(volume);
        seekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {

            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                volume = progress;
                audioManager.setStreamVolume(AudioManager.STREAM_ALARM, progress, 0);
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
                if (ringtone != null && !ringtone.isPlaying()) {
                    ringtone.play();
                }
                if (mediaPlayer != null) {
                    try {
                        if (!mediaPlayer.isPlaying()) {
                            mediaPlayer.prepare();
                            mediaPlayer.start();
                        }
                    } catch (IOException | IllegalStateException e) {
                        e.printStackTrace();
                    }
                }
            }
        });

        return view;
    }

    @Override
    public void onDialogClosed(boolean positiveResult) {
        final AthanVolumePreference athanPref = (AthanVolumePreference) getPreference();
        if (ringtone != null) {
            ringtone.stop();
        }
        if (mediaPlayer != null) {
            try {
                if (mediaPlayer.isPlaying()) {
                    mediaPlayer.stop();
                    mediaPlayer.release();
                }
            } catch (IllegalStateException e) {
                e.printStackTrace();
            }
        }
        if (positiveResult) {
            athanPref.setVolume(volume);
        }
    }
}

File: PersianCalendar/src/main/java/com/byagowi/persiancalendar/view/preferences/AthanVolumeDialog.kt
package com.byagowi.persiancalendar.view.preferences

import android.content.Context
import android.media.AudioManager
import android.media.MediaPlayer
import android.media.Ringtone
import android.media.RingtoneManager
import android.view.View
import android.widget.SeekBar
import androidx.preference.PreferenceDialogFragmentCompat
import com.byagowi.persiancalendar.R
import com.byagowi.persiancalendar.util.UIUtils
import com.byagowi.persiancalendar.util.Utils
import java.io.IOException

class AthanVolumeDialog : PreferenceDialogFragmentCompat() {
  private val TAG = AthanVolumeDialog::class.java.name

  private var volume: Int = 0
  private var audioManager: AudioManager? = null
  private var ringtone: Ringtone? = null
  private var mediaPlayer: MediaPlayer? = null

  override fun onCreateDialogView(context: Context): View {
    val view = super.onCreateDialogView(context)

    val athanPref = preference as AthanVolumePreference

    audioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
    audioManager?.setStreamVolume(AudioManager.STREAM_ALARM, athanPref.volume, 0)

    val customAthanUri = Utils.getCustomAthanUri(context)
    if (customAthanUri != null) {
      ringtone = RingtoneManager.getRingtone(context, customAthanUri)
      ringtone?.streamType = AudioManager.STREAM_ALARM
      ringtone?.play()
    } else {
      val player = MediaPlayer()
      try {
        player.setDataSource(context, UIUtils.getDefaultAthanUri(context))
        player.setAudioStreamType(AudioManager.STREAM_ALARM)
        player.prepare()
      } catch (e: IOException) {
        e.printStackTrace()
      }

      try {
        player.start()
        mediaPlayer = player
      } catch (e: Exception) {
        e.printStackTrace()
      }

    }

    val seekBar = view.findViewById<SeekBar>(R.id.sbVolumeSlider)

    volume = athanPref.volume
    seekBar.progress = volume
    seekBar.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {

      override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
        volume = progress
        audioManager?.setStreamVolume(AudioManager.STREAM_ALARM, progress, 0)
      }

      override fun onStartTrackingTouch(seekBar: SeekBar) {}

      override fun onStopTrackingTouch(seekBar: SeekBar) {
        val rngtone = ringtone
        if (rngtone != null && !rngtone.isPlaying) {
          rngtone.play()
        }
        val player = mediaPlayer
        if (player != null) {
          try {
            if (!player.isPlaying) {
              player.prepare()
              player.start()
            }
          } catch (e: IOException) {
            e.printStackTrace()
          } catch (e: IllegalStateException) {
            e.printStackTrace()
          }

        }
      }
    })

    return view
  }

  override fun onDialogClosed(positiveResult: Boolean) {
    val athanPref = preference as AthanVolumePreference
    ringtone?.stop()
    val player = mediaPlayer
    if (player != null) {
      try {
        if (player.isPlaying) {
          player.stop()
          player.release()
        }
      } catch (e: IllegalStateException) {
        e.printStackTrace()
      }

    }
    if (positiveResult) {
      athanPref.volume = volume
    }
  }
}
